storage_file/models/storage_file.py

```diff
@@ -0,0 +1,230 @@
+            return self.slug
+
+    def _prepare_meta_for_file(self):
+        bin_data = base64.b64decode(self.data)
+        checksum = hashlib.sha1(bin_data).hexdigest()
+        relative_path = self._build_relative_path(checksum)
+        return {
+            "checksum": checksum,
+            "file_size": len(bin_data),
+            "relative_path": relative_path,
+        }
+
+    def _inverse_data(self):
+        for record in self:
+            record.write(record._prepare_meta_for_file())
+            record.backend_id.sudo().add(
+                record.relative_path,
+                record.data,
+                mimetype=record.mimetype,
+                binary=False,
+            )
+
+    def _compute_data(self):
+        for rec in self:
+            if self._context.get("bin_size"):
+                rec.data = rec.file_size
+            elif rec.relative_path:
+                rec.data = rec.backend_id.sudo().get(rec.relative_path, binary=False)
+            else:
+                rec.data = None
+
+    @api.depends("relative_path", "backend_id")
+    def _compute_url(self):
+        for record in self:
+            record.url = record._get_url()
+
+    @api.depends("relative_path", "backend_id")
+    def _compute_url_path(self):
+        # Keep this separated from `url` to avoid multiple compute:
+        # you'll need either one or the other.
+        for record in self:
+            record.url_path = record._get_url(exclude_base_url=True)
+
+    def _get_url(self, exclude_base_url=False):
+        """Retrieve file URL based on backend params.
+
+        :param exclude_base_url: skip base_url
+        """
+        return self.backend_id._get_url_for_file(
+            self, exclude_base_url=exclude_base_url
+        )
+
+    @api.depends("slug")
+    def _compute_internal_url(self):
+        for record in self:
+            record.internal_url = record._get_internal_url()
+
+    def _get_internal_url(self):
+        """Retrieve file URL to load file directly from the storage.
+
+        It is recommended to use this for Odoo backend internal usage
+        to not generate traffic on external services.
+        """
+        return f"/storage.file/{self.slug}"
+
+    @api.depends("name")
+    def _compute_extract_filename(self):
+        for rec in self:
+            if rec.name:
+                rec.filename, rec.extension = os.path.splitext(rec.name)
+                mime, __ = mimetypes.guess_type(rec.name)
+            else:
+                rec.filename = rec.extension = mime = False
+            rec.mimetype = mime
+
+    def unlink(self):
+        if self._context.get("cleanning_storage_file"):
+            super().unlink()
+        else:
+            self.write({"to_delete": True, "active": False})
+        return True
+
+    @api.model
+    def _clean_storage_file(self):
+        # we must be sure that all the changes are into the DB since
+        # we by pass the ORM
+        self.env.flush_all()
```

Comment:
```suggestion
        self.flush_model(["to_delete"])
```
nitpicking: this is actually more accurate